{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import csv\n",
    "#import seaborn as sns\n",
    "path_gff_tes = 'data/results/all.noun.no00.gff'\n",
    "path_gff_genes = 'data/results/Stuberosum_genes_sin_chr0.gff3'\n",
    "path_genome = 'data/Stuberosum_genome.gff3'\n",
    "te_classes = ['TRIM', 'LARD','LTR','SINE', 'LINE','TIR','helitron', 'MITE'] #falta mite"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genome = pd.read_csv(path_genome, sep='\\t', header=None)\n",
    "df_genome.columns = ['seqname' , 'source' , 'feature' , 'start' , 'end' , 'score' , 'strand' , 'frame' , 'attribute']\n",
    "\n",
    "df_tes = pd.read_csv(path_gff_tes, sep='\\t', header=None)\n",
    "df_tes.columns = ['seqname' , 'source' , 'feature' , 'start' , 'end' , 'score' , 'strand' , 'frame' , 'attribute']\n",
    "\n",
    "df_genes = pd.read_csv(path_gff_genes, sep='\\t', header=None)\n",
    "df_genes.columns = ['seqname' , 'source' , 'feature' , 'start' , 'end' , 'score' , 'strand' , 'frame' , 'attribute']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>phytozomev12</td>\n",
       "      <td>gene</td>\n",
       "      <td>71267700</td>\n",
       "      <td>71269327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PGSC0003DMG400000001.v4.03;Name=PGSC0003DMG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>phytozomev12</td>\n",
       "      <td>gene</td>\n",
       "      <td>71271492</td>\n",
       "      <td>71272645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PGSC0003DMG400000002.v4.03;Name=PGSC0003DMG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname        source feature     start       end score strand frame  \\\n",
       "0   chr01  phytozomev12    gene  71267700  71269327     .      +     .   \n",
       "1   chr01  phytozomev12    gene  71271492  71272645     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  ID=PGSC0003DMG400000001.v4.03;Name=PGSC0003DMG...  \n",
       "1  ID=PGSC0003DMG400000002.v4.03;Name=PGSC0003DMG...  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes['gene_id'] = df_genes.attribute.str.split(\";\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>phytozomev12</td>\n",
       "      <td>gene</td>\n",
       "      <td>71267700</td>\n",
       "      <td>71269327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PGSC0003DMG400000001.v4.03;Name=PGSC0003DMG...</td>\n",
       "      <td>ID=PGSC0003DMG400000001.v4.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>phytozomev12</td>\n",
       "      <td>gene</td>\n",
       "      <td>71271492</td>\n",
       "      <td>71272645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=PGSC0003DMG400000002.v4.03;Name=PGSC0003DMG...</td>\n",
       "      <td>ID=PGSC0003DMG400000002.v4.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname        source feature     start       end score strand frame  \\\n",
       "0   chr01  phytozomev12    gene  71267700  71269327     .      +     .   \n",
       "1   chr01  phytozomev12    gene  71271492  71272645     .      +     .   \n",
       "\n",
       "                                           attribute  \\\n",
       "0  ID=PGSC0003DMG400000001.v4.03;Name=PGSC0003DMG...   \n",
       "1  ID=PGSC0003DMG400000002.v4.03;Name=PGSC0003DMG...   \n",
       "\n",
       "                         gene_id  \n",
       "0  ID=PGSC0003DMG400000001.v4.03  \n",
       "1  ID=PGSC0003DMG400000002.v4.03  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genes.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>ST_annotation</td>\n",
       "      <td>TE</td>\n",
       "      <td>23365</td>\n",
       "      <td>24238</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>TRIM_id:1163;source_name:RLX_family19_TRIM_ret...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>ST_annotation</td>\n",
       "      <td>TE</td>\n",
       "      <td>26437</td>\n",
       "      <td>27916</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>TRIM_id:5841;source_name:RLX_family32_TRIM_ret...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname         source feature  start    end score strand frame  \\\n",
       "0   chr01  ST_annotation      TE  23365  24238     .      +     .   \n",
       "1   chr01  ST_annotation      TE  26437  27916     .      +     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  TRIM_id:1163;source_name:RLX_family19_TRIM_ret...  \n",
       "1  TRIM_id:5841;source_name:RLX_family32_TRIM_ret...  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tes.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tes['te_id'] = df_tes.attribute.str.split(\";\").str[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>te_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr01</td>\n",
       "      <td>ST_annotation</td>\n",
       "      <td>TE</td>\n",
       "      <td>23365</td>\n",
       "      <td>24238</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>TRIM_id:1163;source_name:RLX_family19_TRIM_ret...</td>\n",
       "      <td>TRIM_id:1163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr01</td>\n",
       "      <td>ST_annotation</td>\n",
       "      <td>TE</td>\n",
       "      <td>26437</td>\n",
       "      <td>27916</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>TRIM_id:5841;source_name:RLX_family32_TRIM_ret...</td>\n",
       "      <td>TRIM_id:5841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname         source feature  start    end score strand frame  \\\n",
       "0   chr01  ST_annotation      TE  23365  24238     .      +     .   \n",
       "1   chr01  ST_annotation      TE  26437  27916     .      +     .   \n",
       "\n",
       "                                           attribute         te_id  \n",
       "0  TRIM_id:1163;source_name:RLX_family19_TRIM_ret...  TRIM_id:1163  \n",
       "1  TRIM_id:5841;source_name:RLX_family32_TRIM_ret...  TRIM_id:5841  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tes.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37482\n"
     ]
    }
   ],
   "source": [
    "df_genes['new_start'] = df_genes[['start','end']].min(axis=1)\n",
    "df_genes['new_send'] = df_genes[['start','end']].max(axis=1)\n",
    "df_genes['start'] = df_genes['new_start']\n",
    "df_genes['end'] = df_genes['new_send']\n",
    "df_genes = df_genes.drop('new_start',axis=1).drop('new_send',axis=1)\n",
    "df_genes.sort_values(['seqname', 'start'], inplace=True)\n",
    "print(len(df_genes.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load genome structure\n",
    "df_genes_genome = {}\n",
    "for k,chromosome in df_genome.iterrows():\n",
    "    df_genes_genome[chromosome.seqname] = df_genes[(df_genes.seqname == chromosome.seqname)].reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "#te_classes = ['LARD','SINE']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "class: TRIM\n",
      "class: LARD\n",
      "class: LTR\n",
      "class: SINE\n",
      "class: LINE\n",
      "class: TIR\n",
      "class: helitron\n",
      "class: MITE\n"
     ]
    }
   ],
   "source": [
    "distances_all = {}\n",
    "for te_class in te_classes:\n",
    "    print(\"class: %s\" % (te_class,))\n",
    "    #calculate distance to closest gene\n",
    "    df_tes_class = df_tes[df_tes.attribute.str.contains(te_class + \"_id\")]\n",
    "    distances_all[te_class] = []\n",
    "    for k1, te in df_tes_class.iterrows():\n",
    "        te_id = \"%s %s_%s_%s\" % (te.te_id, te.seqname, te.start, te.end)\n",
    "        df_curr = df_genes_genome[te.seqname]\n",
    "        distances = []\n",
    "        distance_prev = False\n",
    "        for k2, gene in df_curr.iterrows():\n",
    "            gene_id = \"%s %s_%s_%s\" % (gene.gene_id, gene.seqname, gene.start, gene.end)\n",
    "            if te.start < gene.start:\n",
    "                first = te\n",
    "                second = gene\n",
    "            else:\n",
    "                first = gene\n",
    "                second = te\n",
    "            distance = max(0, second.start - first.end)\n",
    "            if distance_prev and distance > distance_prev:\n",
    "                break\n",
    "            distances.append((distance,te_id,gene_id))\n",
    "            distance_prev = distance\n",
    "        distances_all[te_class].append(min(distances, key=lambda x:x[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "for te_class in te_classes:\n",
    "    distance = distances_all[te_class]\n",
    "    with open(\"data/results/dist_\" + te_class + \".csv\", \"w\") as f:\n",
    "        writer = csv.writer(f)\n",
    "        writer.writerows(distance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD8CAYAAAB+UHOxAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAD+FJREFUeJzt3X+MZWV9x/H3p7sLWiGC7kQ3sOtgJDbaKNANQjSGaLSABv4oJksaRavZxGoqqUkDmmD0L+0f2ipG3AgVjUUUf3SLayhVErWJq8O6/Fypq6VhCborKEj91bXf/nEfcRhnnbvDuTN793m/kps55znPPef7hct85p5z7yFVhSSpP3+02gVIklaHASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnq1NrVOvD69etrdnZ2tQ4vSVPp1ltv/XFVzQyxr1ULgNnZWebm5lbr8JI0lZL891D78hSQJHXKAJCkThkAktQpA0CSOmUASFKnlgyAJE9K8q0ktyW5K8m7F5lzbJLrk+xNsjPJ7CSKlSQNZ5x3AL8CXlZVLwROA85NctaCOW8EflJVzwE+ALxv2DIlSUNbMgBq5NG2uq49Fv5/JC8Erm3LNwAvT5LBqpQkDW6sawBJ1iTZDewHbq6qnQumnATcB1BVB4GHgacPWagkaVhjfRO4qn4DnJbkBOALSf60qu483IMl2QpsBdi0adPhPv0xs5d9adnPPRz3vvdVK3IcSVoNh/UpoKr6KXALcO6CTfcDGwGSrAWeCjy4yPO3VdXmqto8MzPIrSwkScs0zqeAZtpf/iR5MvAK4LsLpm0HLmnLFwFfraqF1wkkSUeQcU4BbQCuTbKGUWB8pqpuTPIeYK6qtgNXA59Mshd4CNgysYolSYNYMgCq6nbg9EXGr5i3/EvgNcOWJkmaJL8JLEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVNLBkCSjUluSXJ3kruSvG2ROeckeTjJ7va4YjLlSpKGsnaMOQeBt1fVriTHA7cmubmq7l4w7+tV9erhS5QkTcKS7wCq6oGq2tWWfwbsAU6adGGSpMk6rGsASWaB04Gdi2w+O8ltSb6c5PmHeP7WJHNJ5g4cOHDYxUqShjN2ACQ5DvgccGlVPbJg8y7gWVX1QuBDwBcX20dVbauqzVW1eWZmZrk1S5IGMFYAJFnH6Jf/p6rq8wu3V9UjVfVoW94BrEuyftBKJUmDGudTQAGuBvZU1fsPMeeZbR5Jzmz7fXDIQiVJwxrnU0AvBl4L3JFkdxt7B7AJoKquAi4C3pzkIPALYEtV1QTqlSQNZMkAqKpvAFlizpXAlUMVJUmaPL8JLEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdWjIAkmxMckuSu5PcleRti8xJkg8m2Zvk9iRnTKZcSdJQ1o4x5yDw9qraleR44NYkN1fV3fPmnAec2h4vAj7SfkqSjlBLvgOoqgeqaldb/hmwBzhpwbQLgU/UyDeBE5JsGLxaSdJgDusaQJJZ4HRg54JNJwH3zVvfx++HhCTpCDLOKSAAkhwHfA64tKoeWc7BkmwFtgJs2rRpObtQ52Yv+9KKHOfe975qRY5zuFaif3vvx1jvAJKsY/TL/1NV9flFptwPbJy3fnIbe5yq2lZVm6tq88zMzHLqlSQNZJxPAQW4GthTVe8/xLTtwOvap4HOAh6uqgcGrFOSNLBxTgG9GHgtcEeS3W3sHcAmgKq6CtgBnA/sBX4OvGH4UiVJQ1oyAKrqG0CWmFPAW4YqSpI0eX4TWJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqeWDIAk1yTZn+TOQ2w/J8nDSXa3xxXDlylJGtraMeZ8HLgS+MQfmPP1qnr1IBVJklbEku8AquprwEMrUIskaQUNdQ3g7CS3JflykucfalKSrUnmkswdOHBgoENLkpZjiADYBTyrql4IfAj44qEmVtW2qtpcVZtnZmYGOLQkabmecABU1SNV9Whb3gGsS7L+CVcmSZqoJxwASZ6ZJG35zLbPB5/ofiVJk7Xkp4CSXAecA6xPsg94F7AOoKquAi4C3pzkIPALYEtV1cQqliQNYskAqKqLl9h+JaOPiUqSpojfBJakThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkTi0ZAEmuSbI/yZ2H2J4kH0yyN8ntSc4YvkxJ0tDGeQfwceDcP7D9PODU9tgKfOSJlyVJmrQlA6CqvgY89AemXAh8oka+CZyQZMNQBUqSJmOIawAnAffNW9/XxiRJR7C1K3mwJFsZnSZi06ZNK3noo8rsZV+a+DHufe+rJn4MSatriHcA9wMb562f3MZ+T1Vtq6rNVbV5ZmZmgENLkpZriADYDryufRroLODhqnpggP1KkiZoyVNASa4DzgHWJ9kHvAtYB1BVVwE7gPOBvcDPgTdMqlhJ0nCWDICquniJ7QW8ZbCKJEkrwm8CS1KnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6tRYAZDk3CT3JNmb5LJFtr8+yYEku9vjTcOXKkka0tqlJiRZA3wYeAWwD/h2ku1VdfeCqddX1VsnUKMkaQLGeQdwJrC3qn5QVb8GPg1cONmyJEmTNk4AnATcN299Xxtb6C+S3J7khiQbF9tRkq1J5pLMHThwYBnlSpKGMtRF4H8FZqvqBcDNwLWLTaqqbVW1uao2z8zMDHRoSdJyjBMA9wPz/6I/uY09pqoerKpftdWPAX82THmSpEkZJwC+DZya5JQkxwBbgO3zJyTZMG/1AmDPcCVKkiZhyU8BVdXBJG8FbgLWANdU1V1J3gPMVdV24G+SXAAcBB4CXj/BmiVJA1gyAACqagewY8HYFfOWLwcuH7Y0SdIk+U1gSeqUASBJnTIAJKlTBoAkdcoAkKROGQCS1CkDQJI6ZQBIUqcMAEnqlAEgSZ0yACSpUwaAJHXKAJCkThkAktQpA0CSOmUASFKnDABJ6pQBIEmdMgAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASerUWAGQ5Nwk9yTZm+SyRbYfm+T6tn1nktmhC5UkDWvJAEiyBvgwcB7wPODiJM9bMO2NwE+q6jnAB4D3DV2oJGlY47wDOBPYW1U/qKpfA58GLlww50Lg2rZ8A/DyJBmuTEnS0MYJgJOA++at72tji86pqoPAw8DThyhQkjQZa1fyYEm2Alvb6qNJ7lmhQ68Hfny4T8qReSJrWb0crhXqfUV6WY5l9n/E9nM4Wu9HRS/N2L0cof/Nz7ceeNZQOxsnAO4HNs5bP7mNLTZnX5K1wFOBBxfuqKq2AduWV+ryJZmrqs0rfdxJsJcj19HUj70cmVovs0Ptb5xTQN8GTk1ySpJjgC3A9gVztgOXtOWLgK9WVQ1VpCRpeEu+A6iqg0neCtwErAGuqaq7krwHmKuq7cDVwCeT7AUeYhQSkqQj2FjXAKpqB7BjwdgV85Z/Cbxm2NIGteKnnSbIXo5cR1M/9nJkGrSXeKZGkvrkrSAkqVNTGQBJrkmyP8md88aeluTmJN9rP09s40nywXabituTnDHvOZe0+d9Lcslix1qBXjYmuSXJ3UnuSvK2ae0nyZOSfCvJba2Xd7fxU9otQva2W4Yc08YPeQuRJJe38XuS/PlK9zKvjjVJvpPkxrY+zb3cm+SOJLuTzLWxqXudtRpOSHJDku8m2ZPk7Cnu5bnt38lvH48kuXRF+qmqqXsALwXOAO6cN/b3wGVt+TLgfW35fODLQICzgJ1t/GnAD9rPE9vyiavQywbgjLZ8PPCfjG65MXX9tJqOa8vrgJ2txs8AW9r4VcCb2/JfA1e15S3A9W35ecBtwLHAKcD3gTWr9Fr7W+CfgRvb+jT3ci+wfsHY1L3OWh3XAm9qy8cAJ0xrLwv6WgP8kNFn/Sfez6o1OsA/qFkeHwD3ABva8gbgnrb8UeDihfOAi4GPzht/3LxV7OtfgFdMez/AHwO7gBcx+hLO2jZ+NnBTW74JOLstr23zAlwOXD5vX4/NW+EeTga+ArwMuLHVNpW9tGPfy+8HwNS9zhh9z+i/aNcwp7mXRXp7JfAfK9XPVJ4COoRnVNUDbfmHwDPa8qFuZTHOLS5WVDttcDqjv5ynsp92ymQ3sB+4mdFfvD+t0S1CFtZ1qFuIHBG9AP8A/B3wf2396UxvLwAF/FuSWzP6Vj5M5+vsFOAA8E/t9NzHkjyF6exloS3AdW154v0cTQHwmBrF31R9vCnJccDngEur6pH526apn6r6TVWdxuiv5zOBP1nlkpYlyauB/VV162rXMqCXVNUZjO7s+5YkL52/cYpeZ2sZnQL+SFWdDvwPo1Mkj5miXh7TriddAHx24bZJ9XM0BcCPkmwAaD/3t/FD3cpinFtcrIgk6xj98v9UVX2+DU9tPwBV9VPgFkanSU7I6BYhC+t6rOY8/hYiR0IvLwYuSHIvozvgvgz4R6azFwCq6v72cz/wBUYBPY2vs33Avqra2dZvYBQI09jLfOcBu6rqR2194v0cTQEw/3YUlzA6l/7b8de1K+dnAQ+3t1U3Aa9McmK7uv7KNraikoTRN6n3VNX7522aun6SzCQ5oS0/mdG1jD2MguCiNm1hL4vdQmQ7sKV9suYU4FTgWyvTxUhVXV5VJ9fovitbWm1/yRT2ApDkKUmO/+0yo9fHnUzh66yqfgjcl+S5bejlwN1MYS8LXMzvTv/ASvSzmhc8nsCFkuuAB4D/ZfTXwBsZnW/9CvA94N+Bp7W5YfQ/tPk+cAewed5+/grY2x5vWKVeXsLord3twO72OH8a+wFeAHyn9XIncEUbfzajX3p7Gb29PbaNP6mt723bnz1vX+9sPd4DnLfKr7dz+N2ngKayl1b3be1xF/DONj51r7NWw2nAXHutfZHRp16mspdWx1MYvWN86ryxiffjN4ElqVNH0ykgSdJhMAAkqVMGgCR1ygCQpE4ZAJLUKQNAkjplAEhSpwwASerU/wP59BVqdlJbnwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "a = np.histogram(distances_all['MITE'])\n",
    "plt.hist(a, bins='auto')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "distances_kb = {}\n",
    "distances_all = {}\n",
    "distances_kb_filtered = {}\n",
    "for te_class in te_classes:\n",
    "    #calculate distance to closest gene\n",
    "    df_tes_class = df_tes[df_tes.attribute.str.contains(te_class + \"_id\")]\n",
    "    distances = []\n",
    "    for k,v in df_tes_class.iterrows():\n",
    "        midd = (v.start + v.end) / 2\n",
    "        df_curr = df_genes_genome[v.seqname]\n",
    "        df_res = df_curr.loc[(df_curr.start - midd).abs().argsort()[:1]]\n",
    "        if len(df_res.index) > 0:\n",
    "            closest = (int(df_res.start) + int(df_res.end)) / 2\n",
    "            distance = abs(midd - closest)\n",
    "            distances.append(distance)\n",
    "        else:\n",
    "            print(v.seqname, midd)\n",
    "            print(df_res)\n",
    "            print(df_curr)\n",
    "            print('error')\n",
    "    print(te_class, min(distances), max(distances))\n",
    "    distances_kb[te_class] = []\n",
    "    for d in distances:\n",
    "        distances_kb[te_class].append(d/1000)\n",
    "    #! pip3 install seaborn\n",
    "    distances_kb_filtered[te_class] = []\n",
    "    for distance_kb in distances_kb[te_class]:\n",
    "        if distance_kb <= 500:\n",
    "            distances_kb_filtered[te_class].append(distance_kb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for te_class in te_classes:\n",
    "    # seaborn histogram\n",
    "    plt.figure()\n",
    " \n",
    "    sns.distplot(distances_kb_filtered[te_class], hist=True, kde=False, \n",
    "                 bins=80, color = 'blue',\n",
    "                 hist_kws={'edgecolor':'black'},rug=False)\n",
    "    # Add labels\n",
    "    plt.title(te_class + ' distance to closest gene in the genome')\n",
    "    plt.xlabel('Distance to closest gene (kb)')\n",
    "    plt.ylabel('Number of ' + te_class)\n",
    "    plt.savefig('data/results/plots/histogram' + te_class + '.png', dpi=800,bbox_inches='tight')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
